ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Tests/Rollback.hs

```diff
@@ -24,67 +24,100 @@ import           Test.Util.TestBlock (TestBlock, unTestHash)
 -- not* change branches when sent a rollback to a block strictly older than 'k'
 -- blocks before the current selection.
-prop_rollback :: Bool -> QC.Gen QC.Property
-prop_rollback wantRollback = do
+prop_cannotRollback :: QC.Gen QC.Property
+prop_cannotRollback = do
   genesisTest <- genChains 1
 
-  let schedule = rollbackSchedule (gtBlockTree genesisTest)
+  let SecurityParam k = gtSecurityParam genesisTest
+      schedule = rollbackSchedule (fromIntegral (k + 1)) (gtBlockTree genesisTest)
 
-  -- | We consider the test case interesting if we want a rollback and we can
-  -- actually get one, or if we want no rollback and we cannot actually get one.
+  -- We consider the test case interesting if it allows to rollback even if
+  -- the implementation doesn't
   pure $
-    wantRollback == canRollbackFromTrunkTip (gtSecurityParam genesisTest) (gtBlockTree genesisTest)
+    alternativeChainIsLongEnough (gtSecurityParam genesisTest) (gtBlockTree genesisTest)
+      &&
+    honestChainIsLongEnough (gtSecurityParam genesisTest) (gtBlockTree genesisTest)
     ==>
       runSimOrThrow $ runTest schedulerConfig genesisTest schedule $ \StateView{svSelectedChain} ->
         let headOnAlternativeChain = case AF.headHash svSelectedChain of
               GenesisHash    -> False
               BlockHash hash -> any (0 /=) $ unTestHash hash
         in
-        -- The test passes if we want a rollback and we actually end up on the
-        -- alternative chain or if we want no rollback and end up on the trunk.
-        wantRollback == headOnAlternativeChain
+        -- The test passes if we end up on the trunk.
+        not headOnAlternativeChain
 
   where
     schedulerConfig = noTimeoutsSchedulerConfig defaultPointScheduleConfig
 
-    -- A schedule that advertises all the points of the trunk, then switches to
-    -- the first alternative chain of the given block tree.
-    --
-    -- PRECONDITION: Block tree with at least one alternative chain.
-    rollbackSchedule :: BlockTree TestBlock -> PointSchedule
-    rollbackSchedule blockTree =
-      let trunk = btTrunk blockTree
-          branch = btbSuffix $ head $ btBranches blockTree
-          states = banalStates trunk ++ banalStates branch
-          peers = peersOnlyHonest states
-          pointSchedule = balanced peers
-       in fromJust pointSchedule
+-- A schedule that advertises all the points of the trunk up until the kth
+-- block after the intersection, then switches to the first alternative
+-- chain of the given block tree.
+--
+-- PRECONDITION: Block tree with at least one alternative chain.
+rollbackSchedule :: Int -> BlockTree TestBlock -> PointSchedule
+rollbackSchedule k blockTree =
+  let branch = head $ btBranches blockTree
+      trunk = btTrunk blockTree
+      prefixLen = AF.length $ btbPrefix branch
+      trunkPrefix = AF.takeOldest (k + prefixLen) trunk
+      branchSuffix = btbSuffix branch
+      states = banalStates trunkPrefix ++ banalStates branchSuffix
```

Comment:
I think I would prefer to see it written in three parts: the prefix, then `k` blocks on the trunk, then the other branch. Something like (untested):
```
let prefix = btbPrefix branch
    trunkSuffix = AF.takeOlders k $ btbTrunkSuffix branch
    branchSuffix = btbSuffix branch
    states = banalStates prefix ++ banalStates trunkSuffix ++ banalStates branchSuffix
```
I think I understand better what's supposed to happen. What do you think?

Reply:
Fine by me. I'll rewrite.

Reply:
Addressed in 87cc585ba